File: app/src/main/java/ru/goodibunakov/prodvtest/fragments/MainFragment.java
package ru.goodibunakov.prodvtest.fragments;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import ru.goodibunakov.prodvtest.Constants;
import ru.goodibunakov.prodvtest.MainActivity;
import ru.goodibunakov.prodvtest.R;
import ru.goodibunakov.prodvtest.model.List;
import ru.goodibunakov.prodvtest.model.WeatherForecastModel;
import ru.goodibunakov.prodvtest.model.WeatherModel;
import ru.goodibunakov.prodvtest.utils.DateUtils;
import ru.goodibunakov.prodvtest.utils.ImageUtils;
import ru.goodibunakov.prodvtest.utils.TranslateUtils;

public class MainFragment extends Fragment {

    private Unbinder unbinder;
    private String city;
    private int imageId;
    private String[] forecastTexts = new String[5];

    @BindView(R.id.progressbar)
    ProgressBar progressBar;
    @BindView(R.id.city)
    TextView cityText;
    @BindView(R.id.temp)
    TextView tempText;
    @BindView(R.id.image)
    ImageView image;
    @BindView(R.id.sheet_text1)
    TextView sheetText1;
    @BindView(R.id.sheet_text2)
    TextView sheetText2;
    @BindView(R.id.sheet_text3)
    TextView sheetText3;
    @BindView(R.id.sheet_text4)
    TextView sheetText4;
    @BindView(R.id.sheet_text5)
    TextView sheetText5;

    public MainFragment() {
    }


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_main, container, false);
        unbinder = ButterKnife.bind(this, view);

        if (savedInstanceState == null) {
            updateScreen();
        } else {
            setRestoredData(savedInstanceState);
        }

        return view;
    }

    private void setRestoredData(Bundle savedInstanceState) {
        cityText.setText(savedInstanceState.getString(Constants.KEY_CITY));
        tempText.setText(savedInstanceState.getString(Constants.KEY_TEMP));
        if (savedInstanceState.getInt(Constants.KEY_IMAGE_ID) != 0) {
            image.setImageResource(savedInstanceState.getInt(Constants.KEY_IMAGE_ID));
        }

        forecastTexts = savedInstanceState.getStringArray(Constants.KEY_FORECAST_ARRAY);
        if (forecastTexts != null && forecastTexts.length == 5) {
            sheetText1.setText(forecastTexts[0]);
            sheetText2.setText(forecastTexts[1]);
            sheetText3.setText(forecastTexts[2]);
            sheetText4.setText(forecastTexts[3]);
            sheetText5.setText(forecastTexts[4]);
        }
    }

    public void updateScreen() {
        city = TranslateUtils.fromRuToEng();
        getTodayWeather();
        getForecastWeather();
    }

    private void getTodayWeather() {
        progressBar.setVisibility(View.VISIBLE);
        Call<WeatherModel> callToday = ((MainActivity) Objects.requireNonNull(getActivity())).getApiService().getToday(city);
        callToday.enqueue(new Callback<WeatherModel>() {
            @Override
            public void onResponse(@NonNull Call<WeatherModel> call, @NonNull Response<WeatherModel> response) {
                WeatherModel data = response.body();
                progressBar.setVisibility(View.INVISIBLE);

                if (response.isSuccessful()) {
                    if (data != null) {
                        Log.d("debug", "WeatherModel = " + data.toString());
                        cityText.setText(TranslateUtils.fromEngToRu(data.getName()));
                        tempText.setText(getString(R.string.gradus, String.format(Locale.getDefault(), "%.1f", data.getMain().getTemp())));
                        imageId = ImageUtils.getImageDrawable(data.getWeather().get(0).getDescription());
                        image.setImageResource(imageId);
                    }
                }
            }

            @Override
            public void onFailure(@NonNull Call<WeatherModel> call, @NonNull Throwable t) {
                progressBar.setVisibility(View.INVISIBLE);
                Toast.makeText(getActivity(), getResources().getString(R.string.download_error), Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void getForecastWeather() {
        progressBar.setVisibility(View.VISIBLE);
        Call<WeatherForecastModel> callForecast = ((MainActivity) Objects.requireNonNull(getActivity())).getApiService().getForecast(city);
        callForecast.enqueue(new Callback<WeatherForecastModel>() {
            @Override
            public void onResponse(@NonNull Call<WeatherForecastModel> call, @NonNull Response<WeatherForecastModel> response) {
                WeatherForecastModel data = response.body();
                progressBar.setVisibility(View.INVISIBLE);
                Log.d("debug", "response.body() = " + response.body());
                if (response.isSuccessful()) {
                    if (data != null) {
                        fillSheet(data);
                    }
                }
            }

            @Override
            public void onFailure(@NonNull Call<WeatherForecastModel> call, @NonNull Throwable t) {
                progressBar.setVisibility(View.INVISIBLE);
                Toast.makeText(getActivity(), getResources().getString(R.string.download_error), Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void fillSheet(WeatherForecastModel data) {
        // https://api.openweathermap.org/data/2.5/forecast?q=Prague&appid=b849169e8518a852747d3feae7403e88&units=metric&lang=ru
        ArrayList<List> list_ = data.getList();
        ArrayList<String> dates = new ArrayList<>();
        for (List list : list_) {
            String date = DateUtils.convertDate(list.getDtTxt());
            if (!dates.contains(date)) {
                dates.add(date);
            }
        }

        if (dates.size() > 0) {
            for (int i = 0; i < dates.size(); i++) {
                ArrayList<List> forecast = new ArrayList<>();
                String date = dates.get(i);
                for (List list : list_) {
                    if (list.getDtTxt().contains(date)) {
                        forecast.add(list);
                    }
                }

                double avgTemp = 0;
                double avgWind = 0;
                double avgPressure = 0;

                for (List list : forecast) {
                    avgTemp = avgTemp + list.getMain().getTemp();
                    avgWind = avgWind + list.getWind().getSpeed();
                    avgPressure = avgPressure + list.getMain().getPressure();
                }

                avgTemp = avgTemp / forecast.size();
                avgWind = avgWind / forecast.size();
                avgPressure = avgPressure / forecast.size() / 1.333;
                String dateSheet = DateUtils.convertDateForUI(forecast.get(0).getDtTxt());

                if (i <= 4) {
                    forecastTexts[i] = getString(R.string.sheet_line, dateSheet,
                            String.format(Locale.getDefault(), "%.1f", avgTemp),
                            String.format(Locale.getDefault(), "%.1f", avgWind),
                            String.format(Locale.getDefault(), "%.1f", avgPressure)
                    );
                }
            }
        }

        sheetText1.setText(forecastTexts[0]);
        sheetText2.setText(forecastTexts[1]);
        sheetText3.setText(forecastTexts[2]);
        sheetText4.setText(forecastTexts[3]);
        sheetText5.setText(forecastTexts[4]);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(Constants.KEY_CITY, cityText.getText().toString());
        outState.putString(Constants.KEY_TEMP, tempText.getText().toString());
        outState.putInt(Constants.KEY_IMAGE_ID, imageId);
        outState.putStringArray(Constants.KEY_FORECAST_ARRAY, forecastTexts);
    }
}

File: app/src/main/java/ru/goodibunakov/prodvtest/view/fragments/MainFragment.kt
package ru.goodibunakov.prodvtest.view.fragments

import android.os.Bundle
import android.util.Log
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import kotlinx.android.synthetic.main.bottom_sheet.*
import kotlinx.android.synthetic.main.fragment_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import ru.goodibunakov.prodvtest.Constants
import ru.goodibunakov.prodvtest.view.MainActivity
import ru.goodibunakov.prodvtest.R
import ru.goodibunakov.prodvtest.model.List
import ru.goodibunakov.prodvtest.model.WeatherForecastModel
import ru.goodibunakov.prodvtest.model.WeatherModel
import ru.goodibunakov.prodvtest.utils.DateUtils
import ru.goodibunakov.prodvtest.utils.ImageUtils
import ru.goodibunakov.prodvtest.utils.TranslateUtils
import java.util.*


class MainFragment : Fragment() {

    private lateinit var city: String
    private var imageId: Int = 0
    private var forecastTexts: Array<String> = Array(5) { "" }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_main, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (savedInstanceState == null) {
            updateScreen()
        } else {
            setRestoredData(savedInstanceState)
        }
    }

    private fun setRestoredData(savedInstanceState: Bundle) {
        cityText!!.text = savedInstanceState.getString(Constants.KEY_CITY)
        tempText!!.text = savedInstanceState.getString(Constants.KEY_TEMP)
        if (savedInstanceState.getInt(Constants.KEY_IMAGE_ID) != 0) {
            imageView!!.setImageResource(savedInstanceState.getInt(Constants.KEY_IMAGE_ID))
        }

        forecastTexts = savedInstanceState.getStringArray(Constants.KEY_FORECAST_ARRAY)!!
        if (forecastTexts.size == 5) {
            sheetText1!!.text = forecastTexts[0]
            sheetText2!!.text = forecastTexts[1]
            sheetText3!!.text = forecastTexts[2]
            sheetText4!!.text = forecastTexts[3]
            sheetText5!!.text = forecastTexts[4]
        }
    }

    fun updateScreen() {
        city = TranslateUtils.fromRuToEng()
        getTodayWeather()
        getForecastWeather()
    }

    private fun getTodayWeather() {
        progressBar!!.visibility = View.VISIBLE
        val callToday = (activity as MainActivity).apiService.getToday(city)
        callToday.enqueue(object : Callback<WeatherModel> {
            override fun onResponse(call: Call<WeatherModel>, response: Response<WeatherModel>) {
                val data: WeatherModel? = response.body()
                progressBar!!.visibility = View.INVISIBLE

                if (response.isSuccessful) {
                    if (data != null) {
                        Log.d("debug", "WeatherModel = $data")
                        cityText!!.text = TranslateUtils.fromEngToRu(data.name!!)
                        tempText!!.text = getString(R.string.gradus, String.format(Locale.getDefault(), "%.1f", data.main!!.temp))
                        imageId = ImageUtils.getImageDrawable(data.weather!![0].description!!)
                        imageView!!.setImageResource(imageId)
                    }
                }
            }

            override fun onFailure(call: Call<WeatherModel>, t: Throwable) {
                progressBar!!.visibility = View.INVISIBLE
                Toast.makeText(activity, resources.getString(R.string.download_error), Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun getForecastWeather() {
        progressBar!!.visibility = View.VISIBLE
        val callForecast = (activity as MainActivity).apiService.getForecast(city)
        callForecast.enqueue(object : Callback<WeatherForecastModel> {
            override fun onResponse(call: Call<WeatherForecastModel>, response: Response<WeatherForecastModel>) {
                val data = response.body()
                progressBar!!.visibility = View.INVISIBLE
                Log.d("debug", "response.body() = " + response.body()!!)
                if (response.isSuccessful) {
                    if (data != null) {
                        fillSheet(data)
                    }
                }
            }

            override fun onFailure(call: Call<WeatherForecastModel>, t: Throwable) {
                progressBar!!.visibility = View.INVISIBLE
                Toast.makeText(activity, resources.getString(R.string.download_error), Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun fillSheet(data: WeatherForecastModel) {
        // https://api.openweathermap.org/data/2.5/forecast?q=Prague&appid=b849169e8518a852747d3feae7403e88&units=metric&lang=ru
        val listFromData = data.list
        val dates = ArrayList<String>()
        for (list in listFromData!!) {
            val date = DateUtils.convertDate(list.dtTxt!!)
            if (!dates.contains(date)) {
                dates.add(date)
            }
        }

        if (dates.size > 0) {
            for (i in dates.indices) {
                val forecast = ArrayList<List>()
                val date = dates[i]
                for (list in listFromData) {
                    if (list.dtTxt!!.contains(date)) {
                        forecast.add(list)
                    }
                }

                var avgTemp = 0.0
                var avgWind = 0.0
                var avgPressure = 0.0

                for (list in forecast) {
                    avgTemp += list.main?.temp!!
                    avgWind += list.wind?.speed!!
                    avgPressure += list.main!!.pressure!!
                }

                avgTemp /= forecast.size
                avgWind /= forecast.size
                avgPressure = avgPressure / forecast.size.toDouble() / 1.333
                val dateSheet = DateUtils.convertDateForUI(forecast[0].dtTxt!!)

                if (i <= 4) {
                    forecastTexts[i] = getString(R.string.sheet_line, dateSheet,
                            String.format(Locale.getDefault(), "%.1f", avgTemp),
                            String.format(Locale.getDefault(), "%.1f", avgWind),
                            String.format(Locale.getDefault(), "%.1f", avgPressure)
                    )
                }
            }
        }

        sheetText1!!.text = forecastTexts[0]
        sheetText2!!.text = forecastTexts[1]
        sheetText3!!.text = forecastTexts[2]
        sheetText4!!.text = forecastTexts[3]
        sheetText5!!.text = forecastTexts[4]
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(Constants.KEY_CITY, cityText!!.text.toString())
        outState.putString(Constants.KEY_TEMP, tempText!!.text.toString())
        outState.putInt(Constants.KEY_IMAGE_ID, imageId)
        outState.putStringArray(Constants.KEY_FORECAST_ARRAY, forecastTexts)
    }
}